### ISSUE TITLE
[Tablet Splitting] Read count mismatch for Indexes with automatic tablet splitting

### ISSUE THREAD

**Issue by Arjun-yb** [yugabyte/yugabyte-db#10926]
DB Version: **2.11.2.0-b67**
**Steps**:
1. Create Universe with below GFlags(automatic tablet splitting)
 master:
    tablet_split_low_phase_size_threshold_bytes = 13421
    enable_automatic_tablet_splitting = true
    tablet_split_low_phase_shard_count_per_node = 16
tserver:
    yb_num_shards_per_tserver = 1
    
 2. Create table and create Regular/GIN Index (ASC - range partitioned index)
3. Load 100K data(each row should have more data in indexed column)
4. Read data from the table 
        i. With Index scan (index scan = on) and collect the count
        ii. With sequence scan (index scan = off) and collect the count
5. Compare the count - observe the count mismatch

**Note**: Same is working fine without automatic tablet splitting

**Comment by Arjun-yb** [comment]
To reproduce you can use below example:
Connect to ysqlsh
```
create table employees(id int primary key, name text, description text);
create index idx1 on employees(description ASC);
```

Run attached script: 
mv this file to collect_insert_data.py and run `python collect_insert_data.py`
[collect_insert_data.txt](https://github.com/yugabyte/yugabyte-db/files/7862533/collect_insert_data.txt)

run cmd given by the script result(replace expected values)

Connect to ysqlsh
```
SET enable_indexscan = on;
select count(*) from employees where description='<any existing record>';
SET enable_indexscan = off;
select count(*) from employees where description='<same record>';
```

Observe the difference in count

**Comment by ttyusupov** [comment]
I was able to reproduce the issue using the following steps on commit 73e2c667ded0950ca7fc3387894c0544ca774415:
1. `./bin/yb-ctl create --master_flags="enable_automatic_tablet_splitting=true,tablet_split_low_phase_shard_count_per_node=16,tablet_split_low_phase_size_threshold_bytes=13421" --tserver_flags="yb_num_shards_per_tserver=1"`
2. `./bin/ysqlsh` and execute:
```
create table employees(id int primary key, name text, description text);
create index idx1 on employees(description ASC);
```
3. `./collect_insert_data.py`
4. `./bin/ysqlsh -f sql_data.sql`
5. Wait for the data to be loaded
6. `./bin/ysqlsh` and execute:
```
SET enable_indexscan = on;
select count(*) from employees where description='Engineering has existed since ancient times, when humans devised inventions such as the wedge, lever, wheel and pulley, etc. The term engineering is derived from the
word engineer, which itself dates back to the 14th century when an engineer referred to a constructor of military engines. In this context, now obsolete, an engine referred to a military machine i.e., a mechanical contraption used in war Notable examples of the obsolete usage which have survived to the present day are military engineering corps. The word engine itself is of even older origin, ultimately deriving from the Latin ingenium meaning innate quality, especially mental power, hence a clever invention. Later, as the design of civilian structures, such as bridges and buildings, matured as a technical discipline, the term civil engineering entered the lexicon as a way to distinguish between those specializing in the construction of such non-military projects and those involved in the discipline of military engineering. The science of classical mechanics, sometimes called Newtonian mechanics, formed the scientific basis of much of modern engineering. With the rise of engineering as a profession in the 18th century, the term became more narrowly applied to fields in which mathematics and science were applied to these ends. Similarly, in addition to military and civil engineering, the fields then known as the mechanic arts became incorporated into engineering. Canal building was an important engineering work during the early phases of the Industrial Revolution. John Smeaton was the first self-proclaimed civil engineer and is often regarded as the father of civil engineering. He was an English civil engineer responsible for the design of bridges, canals, harbors, and lighthouses. He was also a capable mechanical engineerand an eminentphysicist. Using a model water wheel, Smeaton conducted experiments for seven years, determining ways to increase efficiency. Smeaton introduced iron axles and gears to water wheels. Smeaton also made mechanical improvements to the Newcomen steam engine. Smeaton designed the third Eddystone Lighthouse (1755) where he pioneered the use of hydraulic lime (a form of mortar which will set under water) and developed a technique involving dovetailed blocks of granite in the building of the lighthouse. He is important in the history, rediscovery of, and development of modern cement, because he identified the compositional requirements needed to obtain hydraulicity in lime; work which led ultimately to the invention ofPortland cement. Applied science lead to the development of the steam engine. The sequence of events began with the invention of the barometer and the measurement of atmospheric pressure by Evangelista Torricelli in 1643, demonstration of the force of atmospheric pressure by Otto von Guericke using the Magdeburg hemispheres in 1656, laboratory experiments by Denis Papin, who built experimental model steam engines and demonstrated the use of a piston, which he published in 1707. Edward Somerset, 2nd Marquess of Worcester published a book of 100 inventions containing a method for raising waters similar to a coffee percolator. Samuel Morland, a mathematician and inventor who worked on pumps, left notes at the Vauxhall Ordinance Office on a steam pump design that Thomas Savery read.. In 1698 Savery built a steam pump called The Miners Friend. It employed both vacuum and pressure. Iron merchant Thomas Newcomen, who built the first commercial piston steam engine in 1712, was not known to have any scientific training. The application of steam-powered cast iron blowing cylinders for providing pressurized air for blast furnaces lead to a large increase in iron production in the late 18th century. The higher furnace temperatures made possible with steam-powered blast allowed for the use of more lime in blast furnaces, which enabled the transition from charcoal to coke. These innovations lowered the cost of iron, making horse railways and iron bridges practical.
The puddling process, patented by Henry Cort in 1784 produced large scale quantities of wrought iron. Hot blast, patented by James Beaumont Neilson in 1828, greatly lowered the amount of fuel needed to smelt iron. With the development of the high pressure steam engine, the power to weight ratio of steam engines made practical steamboats and locomotives possible. New steel making processes, such as the Bessemer process and the open hearth furnace, ushered in an area of heavy engineering in the late 19th century.. One of the most famous engineers of the mid 19th century was Isambard Kingdom Brunel, who built railroads, dockyards and steamships. Offshore platform, Gulf of Mexico The Industrial Revolution created a demand for machinery with metal parts, which led to the development of several machine tools. Boring cast iron cylinders with precision was not possible until John Wilkinson invented his
boring machine, which is considered the first machine tool. Other machine tools included the screw cutting lathe, milling machine, turret lathe and the metal planer. Precision machining techniques were developed in the first half of the 19th century. These included the use of gigs to guide the machining tool over the work and fixtures to hold the work in the proper position. Machine tools and machining techniques capable of producing interchangeable
parts lead to large scale factory production by the late 19th century. The United States census of 1850 listed the occupation of engineer for the first time with a count of 2,000.There were fewer than 50 engineering graduates in the U.S. before 1865. In 1870 there were a dozen U.S. mechanical engineering graduates, with that number increasing to 43 per year in 1875. In 1890, there were 6,000 engineers in civil, mining, mechanical and electrical';
SET enable_indexscan = off;
select count(*) from employees where description='Engineering has existed since ancient times, when humans devised inventions such as the wedge, lever, wheel and pulley, etc. The term engineering is derived from the
word engineer, which itself dates back to the 14th century when an engineer referred to a constructor of military engines. In this context, now obsolete, an engine referred to a military machine i.e., a mechanical contraption used in war Notable examples of the obsolete usage which have survived to the present day are military engineering corps. The word engine itself is of even older origin, ultimately deriving from the Latin ingenium meaning innate quality, especially mental power, hence a clever invention. Later, as the design of civilian structures, such as bridges and buildings, matured as a technical discipline, the term civil engineering entered the lexicon as a way to distinguish between those specializing in the construction of such non-military projects and those involved in the discipline of military engineering. The science of classical mechanics, sometimes called Newtonian mechanics, formed the scientific basis of much of modern engineering. With the rise of engineering as a profession in the 18th century, the term became more narrowly applied to fields in which mathematics and science were applied to these ends. Similarly, in addition to military and civil engineering, the fields then known as the mechanic arts became incorporated into engineering. Canal building was an important engineering work during the early phases of the Industrial Revolution. John Smeaton was the first self-proclaimed civil engineer and is often regarded as the father of civil engineering. He was an English civil engineer responsible for the design of bridges, canals, harbors, and lighthouses. He was also a capable mechanical engineerand an eminentphysicist. Using a model water wheel, Smeaton conducted experiments for seven years, determining ways to increase efficiency. Smeaton introduced iron axles and gears to water wheels. Smeaton also made mechanical improvements to the Newcomen steam engine. Smeaton designed the third Eddystone Lighthouse (1755) where he pioneered the use of hydraulic lime (a form of mortar which will set under water) and developed a technique involving dovetailed blocks of granite in the building of the lighthouse. He is important in the history, rediscovery of, and development of modern cement, because he identified the compositional requirements needed to obtain hydraulicity in lime; work which led ultimately to the invention ofPortland cement. Applied science lead to the development of the steam engine. The sequence of events began with the invention of the barometer and the measurement of atmospheric pressure by Evangelista Torricelli in 1643, demonstration of the force of atmospheric pressure by Otto von Guericke using the Magdeburg hemispheres in 1656, laboratory experiments by Denis Papin, who built experimental model steam engines and demonstrated the use of a piston, which he published in 1707. Edward Somerset, 2nd Marquess of Worcester published a book of 100 inventions containing a method for raising waters similar to a coffee percolator. Samuel Morland, a mathematician and inventor who worked on pumps, left notes at the Vauxhall Ordinance Office on a steam pump design that Thomas Savery read.. In 1698 Savery built a steam pump called The Miners Friend. It employed both vacuum and pressure. Iron merchant Thomas Newcomen, who built the first commercial piston steam engine in 1712, was not known to have any scientific training. The application of steam-powered cast iron blowing cylinders for providing pressurized air for blast furnaces lead to a large increase in iron production in the late 18th century. The higher furnace temperatures made possible with steam-powered blast allowed for the use of more lime in blast furnaces, which enabled the transition from charcoal to coke. These innovations lowered the cost of iron, making horse railways and iron bridges practical.
The puddling process, patented by Henry Cort in 1784 produced large scale quantities of wrought iron. Hot blast, patented by James Beaumont Neilson in 1828, greatly lowered the amount of fuel needed to smelt iron. With the development of the high pressure steam engine, the power to weight ratio of steam engines made practical steamboats and locomotives possible. New steel making processes, such as the Bessemer process and the open hearth furnace, ushered in an area of heavy engineering in the late 19th century.. One of the most famous engineers of the mid 19th century was Isambard Kingdom Brunel, who built railroads, dockyards and steamships. Offshore platform, Gulf of Mexico The Industrial Revolution created a demand for machinery with metal parts, which led to the development of several machine tools. Boring cast iron cylinders with precision was not possible until John Wilkinson invented his
boring machine, which is considered the first machine tool. Other machine tools included the screw cutting lathe, milling machine, turret lathe and the metal planer. Precision machining techniques were developed in the first half of the 19th century. These included the use of gigs to guide the machining tool over the work and fixtures to hold the work in the proper position. Machine tools and machining techniques capable of producing interchangeable
parts lead to large scale factory production by the late 19th century. The United States census of 1850 listed the occupation of engineer for the first time with a count of 2,000.There were fewer than 50 engineering graduates in the U.S. before 1865. In 1870 there were a dozen U.S. mechanical engineering graduates, with that number increasing to 43 per year in 1875. In 1890, there were 6,000 engineers in civil, mining, mechanical and electrical';
```

**Expected result:** counts are the same.
**Actual result:** counts are differents (2776 vs 10000).

**Comment by bmatican** [comment]
Migrating some discussion from Slack.

Seems the problem here is that indexes have an automatic partitioning column added to them, `either ybidxbasectid for non-unique indexes or ybuniqueidxkeysuffix for unique indexes.`

This means that currently when we split, we could split at a boundary such that all the prefix columns are the same across the children, but this last internal column differs. The upper YSQL layer currently does not handle that properly on a scan. However, external tooling, such as ysql_dump, also do not handle that correctly and we do not (yet?) want to expose this column to users, when they create indexes and specify split points manually.

As such, the short term decision was to update the split code for indexes, to not include this last column, in the computation of the middle key. We should make sure to test for both unique and non-unique indexes!
cc @arybochkin @rthallamko3

**Comment by arybochkin** [comment]
The issues is fixed in context of #11463 by commit a2aa0d3140 (index scan was fixed by adjusting range upper bound).
However, several index related issue have been found during the investigation: #12189, #12190 and #12191.
